### PR TITLE
Use `Chunk` builder instead of `Buffer` in `ChunkWriter`

### PR DESCRIPTION
**Pull request by danicheg**
* Throughput before changes:
```
Benchmark           (size)   Mode  Cnt      Score     Error   Units
appendSingleString     100  thrpt    5  21439.685 ± 814.071  ops/ms
appendSingleString    1000  thrpt    5   7990.988 ± 311.376  ops/ms
appendSingleString   10000  thrpt    5    762.789 ±  65.330  ops/ms
appendStringBatch      100  thrpt    5    366.361 ±  21.818  ops/ms
appendStringBatch     1000  thrpt    5     36.774 ±   0.392  ops/ms
appendStringBatch    10000  thrpt    5      3.120 ±   0.066  ops/ms
```
* Memory consumption before changes:
```
appendSingleString:·gc.alloc.rate.norm        100  avgt    5      464.018 ±     0.001    B/op
appendSingleString:·gc.alloc.rate.norm       1000  avgt    5     2272.061 ±     0.001    B/op
appendSingleString:·gc.alloc.rate.norm      10000  avgt    5    20272.914 ±     0.011    B/op
appendStringBatch:·gc.alloc.rate.norm         100  avgt    5    15145.028 ±     0.040    B/op
appendStringBatch:·gc.alloc.rate.norm        1000  avgt    5   148390.629 ±     0.449    B/op
appendStringBatch:·gc.alloc.rate.norm       10000  avgt    5  1531428.925 ±     4.025    B/op
```
* Throughput after changes:
```
Benchmark           (size)   Mode  Cnt      Score      Error   Units
appendSingleString     100  thrpt    5  74746.042 ± 1166.192  ops/ms
appendSingleString    1000  thrpt    5  19464.679 ± 2386.691  ops/ms
appendSingleString   10000  thrpt    5   1329.216 ±   22.597  ops/ms
appendStringBatch      100  thrpt    5    560.002 ±   10.027  ops/ms
appendStringBatch     1000  thrpt    5     54.431 ±    1.881  ops/ms
appendStringBatch    10000  thrpt    5      4.760 ±    0.254  ops/ms
```
* Memory consumption after changes:
```
Benchmark                                  (size)  Mode  Cnt        Score       Error   Units
appendSingleString:·gc.alloc.rate.norm        100  avgt    5      168.005 ±     0.001    B/op
appendSingleString:·gc.alloc.rate.norm       1000  avgt    5     1064.026 ±     0.001    B/op
appendSingleString:·gc.alloc.rate.norm      10000  avgt    5    10064.418 ±     0.002    B/op
appendStringBatch:·gc.alloc.rate.norm         100  avgt    5    16881.091 ±     0.022    B/op
appendStringBatch:·gc.alloc.rate.norm        1000  avgt    5   168085.875 ±     6.935    B/op
appendStringBatch:·gc.alloc.rate.norm       10000  avgt    5  1680145.068 ±     1.331    B/op
```

### TLDR:
* throughput of appending single string increases up to `1.74-3,48x` times, and appending of a batch of strings up to `1.5x` times;
* memory consumption reduces for appending a single string up to `2-2.74`x times, but it increases up to `1.1x` times for appending a batch of strings. That's interesting, but considering the pros shouldn't be a thing.

### CODE REVIEW

**Review by satorg — APPROVED** [review]
LGTM, thanks! (Moreover I personally find it really reasonable to use Chunk builder for building chunks :)

**Comment by rossabaker** [comment]
Is this used anywhere other than Multipart?  Previous benchmarks have shown that writing to a StringBuilder and encoding that to ISO-8859-1 in one pass is faster than anything we've managed by going binary more incrementally, but it's very wasteful of memory.  I'm curious how those compare now.  This is _potentially_ a win for both Ember and Blaze.

Optimizing these encoders is a goal of my ongoing experimentation with HTTP Codecs.  (Defining things bidirectionally from one source is the other.)  This may end up being a useful target for that as well.

**Comment by danicheg** [comment]
Yes, `ChunkWriter` is used by `MultipartEncoder` in http4s, and maybe is used in some satellite projects.

**Comment by danicheg** [comment]
Yes, `ChunkWriter` is used by `MultipartEncoder` in http4s, and maybe is used in some satellite projects.

**Comment by danicheg** [comment]
Whenever this will be rewritten from scratch, an added benchmark can be helpful as a starting point for seeking differences in numbers. Anyway, I'm excited about the current results, and if you don't mind, I merge this even though for a while.
